### PR TITLE
WGSL textureSample tests for 3D and cubemaps

### PR DESCRIPTION
**Pull request by greggman**
Note: 3D compressed textures are not tested as they are not yet supported in any implementation.

Also, see internal comments on cube texture corners.

This is not passing on Vulkan Android and Vulkan linux. In both it's crashing in the driver.

<hr>

**Requirements for PR author:**

- [X] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [X] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [X] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)
- [X] Test have be tested with compatibility mode validation enabled and behave as expected. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [x] Tests are properly located in the test tree.
- [x] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [x] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [x] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts [hunk 1]
```diff
@@ -1167,29 +1418,418 @@ export function generateSamplePoints(
...
+  ];
+}
+
+function wrapFaceCoordToCubeFaceAtEdgeBoundaries(textureSize: number, faceCoord: vec3) {
+  // // convert faceCoord from texels to normalized coords
+  // const uvw = faceCoord.map((v, i) => v / textureSize[i]) as vec3;
+  // // convert to cube coord and back, this may switch faces, wrapping the coord to a different face.
+  // const cubeCoord = convertNormalized3DTexCoordToCubeCoord(uvw);
+  // const coord = convertCubeCoordToNormalized3DTextureCoord(cubeCoord);
+  // // convert back to texels.
```
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts:1648]: commented-out code?

File: src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts [hunk 2]
```diff
@@ -1167,29 +1418,418 @@ export function generateSamplePoints(
...
+        const cosPhi = Math.cos(phi);
+        const ux = cosTheta * sinPhi;
+        const uy = cosPhi;
+        const uz = sinTheta * sinPhi;
+        out.push([ux * r, uy * r, uz * r]);
+      }
+      break;
+    }
+    case 'cube-edges': {
+      /* eslint-disable prettier/prettier */
```
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts:1734]: nit: same here

File: src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts [hunk 3]
```diff
@@ -1167,29 +1418,418 @@ export function generateSamplePoints(
...
+    }) as vec3;
+    return convertNormalized3DTexCoordToCubeCoord(quantizedUVW);
+  });
+}
+
+function wgslTypeFor(data: number | Dimensionality, type: 'f' | 'i' | 'u'): string {
+  if (Array.isArray(data)) {
     switch (data.length) {
+      case 1:
+        return `${type}f32`;
```
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts:1810]: seems to produce `ff32`/`if32`/`uf32`

File: src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts [hunk 4]
```diff
@@ -1167,29 +1418,418 @@ export function generateSamplePoints(
...
       case 2:
         return `vec2${type}`;
       case 3:
         return `vec3${type}`;
+      default:
+        unreachable();
     }
   }
-  return '${type}32';
+  return `${type}f32`;
```
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts:1819]: same

**Review by kainino0x — APPROVED** [review]

File: src/webgpu/shader/execution/expression/call/builtin/textureSample.spec.ts [hunk 5]
```diff
@@ -237,13 +283,107 @@ Parameters:
...
+      .filter(t => {
+        const type = kTextureFormatInfo[t.format].color?.type;
+        const canPotentialFilter = type === 'float' || type === 'unfilterable-float';
+        // We can't easily put random bytes into compressed textures if they are float formats
+        // since we want the range to be +/- 1000 and not +/- infinity or NaN.
+        const isFillable = !isCompressedTextureFormat(t.format) || !t.format.endsWith('float');
+        return canPotentialFilter && isFillable;
+      })
+      .combine('viewDimension', ['3d', 'cube'] as const)
+      // MAINTENANCE_TODO: remove this once 3d compressed textures are allowed.
```
[the remark this answers is not in the document]
Comment (greggman) [reply src/webgpu/shader/execution/expression/call/builtin/textureSample.spec.ts:296]: moved

File: src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts (hunk 1 above)
Comment (greggman) [reply src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts:1646]: doh!, removed

File: src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts (hunk 3 above)
Comment (greggman) [reply src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts:1808]: fixed (that path isn't taken yet but good catch!)

File: src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts (hunk 4 above)
Comment (greggman) [reply src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts:1817]: fixed

File: src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts (hunk 2 above)
Comment (greggman) [reply src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts:1732]: done

**Review by kainino0x — APPROVED** [review]

**Pull request merged**